### ISSUE TITLE
Coverity: CID-153373: possible resource leak in commit_apply_fence

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#869]
I didn't look a this Coverity report too closely, and it may be a false positive or highly unlikely event, but I thought perhaps @garlick could take a quick look and verify:

```C
static void commit_apply_fence (fence_t *f)
495{
496    ctx_t *ctx = f->ctx;
497    wait_t *wait = NULL;
498    int count;
499
   	1. Condition f->errnum, taking false branch.
500    if (f->errnum)
501        goto done;
   	2. alloc_fn: Storage is returned from allocation function wait_create. [show details]
   	3. var_assign: Assigning: wait = storage returned from wait_create((wait_cb_f)commit_apply_fence, f).
   	4. Condition !(wait = wait_create((wait_cb_f)commit_apply_fence, f)), taking false branch.
502    if (!(wait = wait_create ((wait_cb_f)commit_apply_fence, f))) {
503        f->errnum = errno;
504        goto done;
505    }
506
507    /* Make a copy of the root directory.
508     */
   	5. Condition !f->rootcpy_stored, taking true branch.
   	6. Condition !f->rootcpy, taking false branch.
509    if (!f->rootcpy_stored && !f->rootcpy) {
510        json_object *rootdir;
511        if (!load (ctx, ctx->rootdir, wait, &rootdir))
512            goto stall;
513        f->rootcpy = copydir (rootdir);
514    }
515
516    /* Apply each op (e.g. key = val) in sequence to the root copy.
517     * A side effect of walking key paths is to convert DIRREFs to DIRVALs
518     * in the copy. This allows the commit to be self-contained in the rootcpy
519     * until it is unrolled later on.
520     */
   	7. Condition f->ops, taking true branch.
   	8. Condition !f->rootcpy_stored, taking true branch.
521    if (f->ops && !f->rootcpy_stored) {
522        int i, len = json_object_array_length (f->ops);
523        json_object *op, *dirent;
524        const char *key;
525
   	9. Condition i < len, taking true branch.
   	12. Condition i < len, taking false branch.
526        for (i = 0; i < len; i++) {
   	10. Condition !(op = json_c_object_array_get_idx(f->ops, i)), taking true branch.
527            if (!(op = json_object_array_get_idx (f->ops, i))
528                    || !Jget_str (op, "key", &key))
   	11. Continuing loop.
529                continue;
530            dirent = NULL;
531            (void)Jget_obj (op, "dirent", &dirent); /* NULL for unlink */
532            if (commit_link_dirent (ctx, f->rootcpy, key, dirent, wait) < 0) {
533                f->errnum = errno;
534                break;
535            }
536        }
   	13. noescape: Resource wait is not freed or pointed-to in wait_get_usecount. [show details]
   	14. Condition wait_get_usecount(wait) > 0, taking true branch.
537        if (wait_get_usecount (wait) > 0)
   	15. Jumping to label stall.
538            goto stall;
539        if (f->errnum != 0)
540            goto done;
541    }
542
543    /* Unroll the root copy.
544     * When a DIRVAL is found, store an object and replace it with a DIRREF.
545     * Finally, store the unrolled root copy as an object and keep its
546     * reference in f->newroot.  Flushes to content cache are asynchronous
547     * but we don't proceed until they are completed.
548     */
549    if (!f->rootcpy_stored) {
550        if (commit_unroll (ctx, f->rootcpy, wait) < 0)
551            f->errnum = errno;
552        else if (store (ctx, f->rootcpy, f->newroot, wait) < 0)
553            f->errnum = errno;
554        f->rootcpy_stored = true; /* cache takes ownership of rootcpy */
555        f->rootcpy = NULL;
556        if (wait_get_usecount (wait) > 0)
557            goto stall;
558    }
559
560    /* This is the transaction that finalizes the commit by replacing
561     * ctx->rootdir with f->newroot, incrementing the root seq,
562     * and sending out the setroot event for "eventual consistency"
563     * of other nodes.
564     */
565done:
566    if (f->errnum == 0) {
567        if (Jget_ar_len (f->names, &count) && count > 1) {
568            int opcount = 0;
569            (void)Jget_ar_len (f->ops, &opcount);
570            flux_log (ctx->h, LOG_DEBUG, "aggregated %d commits (%d ops)",
571                      count, opcount);
572        }
573        setroot (ctx, f->newroot, ctx->rootseq + 1);
574        setroot_event_send (ctx, f->names);
575    } else {
576        flux_log (ctx->h, LOG_ERR, "commit failed: %s",
577                  flux_strerror (f->errnum));
578        error_event_send (ctx, f->names, f->errnum);
579    }
580    wait_destroy (wait);
581
582    /* Completed: remove from 'ready' list.
583     * N.B. fence_t remains in the fences hash until event is received.
584     */
585    zlist_remove (ctx->ready, f);
586    return;
587
588stall:
589    f->blocked = 1;
   	
CID 153373 (#1 of 1): Resource leak (RESOURCE_LEAK)
16. leaked_storage: Variable wait going out of scope leaks the storage it points to.
590    return;
591}
```

**Comment by garlick** [comment]
This looks to me like a false positive.

The `stall:` label is only entered when `wait` has been placed on a wait queue.  I traced through each code path and couldn't find any situation when this wasn't the case.   Ownership transfers to the waitqueue (e.g. cache entries) until the wait conditions are met and then this function is restarted and eventually it exits via `done:` and frees `wait`.

For an extra check I added an assertion and ran `make check`
```diff
diff --git a/src/modules/kvs/kvs.c b/src/modules/kvs/kvs.c
index d60b814..1d2be92 100644
--- a/src/modules/kvs/kvs.c
+++ b/src/modules/kvs/kvs.c
@@ -586,6 +586,8 @@ done:
     return;
 
 stall:
+    assert (wait != NULL);
+    assert (wait_get_usecount (wait) > 0); /* CID-153373 leak false positive */
     f->blocked = 1;
     return;
 }
```
no failures.

**Comment by grondo** [comment]
Nice, thanks! I'll update Coverity
